File: Sources/DebridStreamer/Services/Debrid/RealDebridService.swift
import Foundation

/// Real-Debrid API integration.
/// API docs: https://api.real-debrid.com/
actor RealDebridService: DebridServiceProtocol {
    let serviceType = DebridServiceType.realDebrid
    private let apiToken: String
    private let baseURL = "https://api.real-debrid.com/rest/1.0"
    private let session: URLSession

    // Reused across calls instead of allocating per `getAccountInfo`.
    private static let expirationFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    init(apiToken: String, session: URLSession = AppHTTP.api) {
        self.apiToken = apiToken
        self.session = session
    }

    // MARK: - Cache Check

    func checkCache(hashes: [String]) async throws -> [String: CacheStatus] {
        guard !hashes.isEmpty else { return [:] }

        // RD has disabled /torrents/instantAvailability (error_code 37).
        // Instead, we return .unknown for all hashes — the resolve flow
        // handles both cached and uncached torrents correctly.
        // Cached torrents become "downloaded" almost instantly after addMagnet + selectFiles.
        var results: [String: CacheStatus] = [:]
        for hash in hashes {
            results[hash.lowercased()] = .unknown
        }
        return results
    }

    // MARK: - Magnet Operations

    func addMagnet(hash: String) async throws -> String {
        let magnet = "magnet:?xt=urn:btih:\(hash)"
        let encodedMagnet = magnet.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? magnet
        let body = "magnet=\(encodedMagnet)"

        // Retry with exponential backoff for transient server errors
        let maxRetries = 5
        var lastError: Error?
        for attempt in 0..<maxRetries {
            do {
                let data = try await requestRaw(path: "/torrents/addMagnet", method: "POST", body: body)

                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let id = json["id"] as? String {
                    return id
                }
                throw DebridError.downloadFailed("Failed to parse magnet response")
            } catch let error as DebridError {
                lastError = error
                if case .httpError(let code, _) = error, (500...599).contains(code) {
                    let delay = UInt64(pow(2.0, Double(attempt))) * 1_000_000_000
                    #if DEBUG
                    print("[RealDebrid] addMagnet HTTP \(code), attempt \(attempt + 1)/\(maxRetries), retrying...")
                    #endif
                    try await Task.sleep(nanoseconds: delay)
                    continue
                }
                throw error
            } catch {
                lastError = error
                break
            }
        }
        throw lastError ?? DebridError.downloadFailed("Failed to add magnet after \(maxRetries) retries")
    }

    func selectFiles(torrentId: String, fileIds: [Int]) async throws {
        let files = fileIds.isEmpty ? "all" : fileIds.map(String.init).joined(separator: ",")
        let body = "files=\(files)"

        // selectFiles returns 204 No Content on success; requestRaw already maps 204
        // to empty Data and returns it, so a successful call never throws. Any non-2xx
        // status is a real failure and must surface — don't swallow it.
        _ = try await requestRaw(path: "/torrents/selectFiles/\(torrentId)", method: "POST", body: body)
    }

    func getStreamURL(torrentId: String) async throws -> StreamInfo {
        // Poll for torrent status — cached torrents become "downloaded" almost instantly
        var status = ""
        var json: [String: Any] = [:]
        // Capped backoff between polls (seconds): start tight so cached torrents
        // resolve fast, then back off and cap at ~5s. Total wait (~21.7s for an
        // uncached torrent) stays comparable to the previous fixed 1s × ~19 schedule,
        // but with far fewer requests while waiting on slow downloads.
        let backoffSchedule: [Double] = [0.4, 0.8, 1.5, 3.0]
        let backoffCap = 5.0
        let maxAttempts = 20

        for attempt in 0..<maxAttempts {
            let data = try await requestRaw(path: "/torrents/info/\(torrentId)", method: "GET")

            guard let parsed = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw DebridError.torrentNotFound(torrentId)
            }
            json = parsed
            status = json["status"] as? String ?? ""

            #if DEBUG
            print("[RealDebrid] Torrent \(torrentId) status: \(status) (attempt \(attempt + 1)/\(maxAttempts))")
            #endif

            if status == "downloaded" {
                break
            }

            if status == "error" || status == "dead" || status == "virus" || status == "magnet_error" {
                throw DebridError.downloadFailed("Torrent status: \(status)")
            }

            // If waiting for file selection, select all files
            if status == "waiting_files_selection" {
                try await selectFiles(torrentId: torrentId, fileIds: [])
            }

            // Wait before next poll, using the capped backoff schedule.
            if attempt < maxAttempts - 1 {
                let delaySeconds = attempt < backoffSchedule.count ? backoffSchedule[attempt] : backoffCap
                try await Task.sleep(nanoseconds: UInt64(delaySeconds * 1_000_000_000))
            }
        }

        guard status == "downloaded" else {
            throw DebridError.downloadFailed("Torrent not ready. Status: \(status)")
        }

        guard let links = json["links"] as? [String], !links.isEmpty else {
            throw DebridError.noFilesAvailable
        }

        let candidates = fileCandidates(from: json, links: links)
        guard let selected = DebridFileSelector.selectBest(from: candidates) else {
            throw DebridError.noFilesAvailable
        }

        // Unrestrict the chosen link to get a direct stream URL
        let streamURL = try await unrestrict(link: selected.link)

        let filename = URL(fileURLWithPath: selected.fileName).lastPathComponent
        let bytes = selected.sizeBytes

        return StreamInfo(
            streamURL: streamURL.absoluteString,
            quality: VideoQuality.parse(from: filename),
            codec: VideoCodec.parse(from: filename),
            audio: AudioFormat.parse(from: filename),
            source: SourceType.parse(from: filename),
            sizeBytes: bytes,
            fileName: filename,
            debridService: "RD"
        )
    }

    private func fileCandidates(from json: [String: Any], links: [String]) -> [DebridFileCandidate] {
        if let files = json["files"] as? [[String: Any]] {
            let selectedFiles = files
                .filter { int64Value($0["selected"]) == 1 }
                .sorted { (lhs, rhs) in
                    int64Value(lhs["id"]) ?? Int64.max < int64Value(rhs["id"]) ?? Int64.max
                }

            var candidates: [DebridFileCandidate] = []
            for (index, link) in links.enumerated() where index < selectedFiles.count {
                let file = selectedFiles[index]
                let path = (file["path"] as? String) ?? (file["filename"] as? String) ?? "Unknown"
                let size = int64Value(file["bytes"]) ?? 0
                candidates.append(
                    DebridFileCandidate(link: link, fileName: path, sizeBytes: size)
                )
            }
            if !candidates.isEmpty {
                return candidates
            }
        }

        let fallbackName = json["filename"] as? String ?? "Unknown"
        let fallbackSize = int64Value(json["bytes"]) ?? 0
        return links.map {
            DebridFileCandidate(link: $0, fileName: fallbackName, sizeBytes: fallbackSize)
        }
    }

    // MARK: - Unrestrict

    func unrestrict(link: String) async throws -> URL {
        let encodedLink = link.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? link
        let body = "link=\(encodedLink)"

        // Retry for transient server errors
        let maxRetries = 5
        var lastError: Error?
        for attempt in 0..<maxRetries {
            do {
                let data = try await requestRaw(path: "/unrestrict/link", method: "POST", body: body)

                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let downloadStr = json["download"] as? String,
                   let downloadURL = URL(string: downloadStr) {
                    return downloadURL
                }
                throw DebridError.downloadFailed("Failed to parse unrestrict response")
            } catch let error as DebridError {
                lastError = error
                if case .httpError(let code, _) = error, (500...599).contains(code) {
                    let delay = UInt64(pow(2.0, Double(attempt))) * 1_000_000_000
                    #if DEBUG
                    print("[RealDebrid] unrestrict HTTP \(code), attempt \(attempt + 1)/\(maxRetries), retrying...")
                    #endif
                    try await Task.sleep(nanoseconds: delay)
                    continue
                }
                throw error
            } catch {
                lastError = error
                break
            }
        }
        throw lastError ?? DebridError.downloadFailed("Failed to unrestrict link after \(maxRetries) retries")
    }

    // MARK: - User Torrents

    /// Check if a hash already exists in the user's torrent list.
    /// Returns the torrent ID if found and status is "downloaded", nil otherwise.
    func findExistingTorrent(hash: String) async throws -> String? {
        // Bound the list instead of pulling the user's entire torrent history every
        // resolve. RD returns torrents newest-first, and a just-resolved torrent is
        // among the most recent, so the first page comfortably covers the realistic
        // "is this already here" case. A miss only costs a redundant addMagnet (which
        // RD dedups server-side), so correctness holds — this is purely an
        // optimization on the GET payload size.
        let data = try await requestRaw(path: "/torrents?limit=100&page=1", method: "GET")
        guard let torrents = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return nil
        }

        let lowerHash = hash.lowercased()
        for torrent in torrents {
            if let torrentHash = torrent["hash"] as? String,
               torrentHash.lowercased() == lowerHash,
               let id = torrent["id"] as? String {
                let status = torrent["status"] as? String ?? ""
                // Only return if already downloaded or has links
                if status == "downloaded" {
                    return id
                }
                // If it's in an error/dead state, delete it so we can re-add
                if status == "error" || status == "dead" || status == "magnet_error" {
                    try? await deleteTorrent(id: id)
                    return nil
                }
                // If it's in progress, return it so we can poll
                return id
            }
        }
        return nil
    }

    /// Delete a torrent from the user's list.
    func deleteTorrent(id: String) async throws {
        _ = try await requestRaw(path: "/torrents/delete/\(id)", method: "DELETE")
    }

    // MARK: - Account

    func validateToken() async throws -> Bool {
        do {
            _ = try await getAccountInfo()
            return true
        } catch {
            return false
        }
    }

    func getAccountInfo() async throws -> DebridAccountInfo {
        let data = try await requestRaw(path: "/user", method: "GET")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DebridError.invalidToken
        }

        let username = json["username"] as? String ?? "Unknown"
        let email = json["email"] as? String
        let premium = json["premium"] as? Int ?? 0
        let points = json["points"] as? Int

        var premiumExpiry: Date?
        if let expirationStr = json["expiration"] as? String {
            premiumExpiry = Self.expirationFormatter.date(from: expirationStr)
        }

        return DebridAccountInfo(
            username: username,
            email: email,
            premiumExpiry: premiumExpiry,
            isPremium: premium > 0,
            points: points
        )
    }

    // MARK: - HTTP

    private func requestRaw(path: String, method: String, body: String? = nil) async throws -> Data {
        guard let url = URL(string: baseURL + path) else {
            throw DebridError.networkError("Invalid URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(apiToken)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 30

        if let body = body {
            request.httpBody = body.data(using: .utf8)
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DebridError.networkError("Invalid response")
        }

        // 204 No Content is success (used by selectFiles, delete)
        if httpResponse.statusCode == 204 {
            return Data()
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw DebridError.invalidToken
            }
            if httpResponse.statusCode == 403 {
                throw DebridError.expired
            }
            if httpResponse.statusCode == 429 {
                throw DebridError.rateLimited
            }
            let errorMsg = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw DebridError.httpError(httpResponse.statusCode, errorMsg)
        }

        return data
    }

    private func int64Value(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber {
            return number.int64Value
        }
        if let intValue = value as? Int {
            return Int64(intValue)
        }
        if let stringValue = value as? String {
            return Int64(stringValue)
        }
        return nil
    }
}

// MARK: - Array Chunking Helper

extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}

File: web/src/services/debrid/RealDebridService.ts
// Port of Sources/DebridStreamer/Services/Debrid/RealDebridService.swift.
//
// A fetch-based Real-Debrid client. Mirrors the Swift actor's HTTP+JSON logic:
// the same API paths, Bearer auth (token never leaked into the query), the
// checkCache "all unknown" behavior, addMagnet/selectFiles/unrestrict parsing,
// the getStreamURL poll-and-pair-files flow, findExistingTorrent, and the
// status->DebridError mapping. The retry/poll SLEEP loops are preserved but the
// delay is injectable (`sleep`, default no-op) so tests never actually wait.

import {
  CacheStatus,
  type DebridAccountInfo,
  type DebridFileCandidate,
  DebridFileSelector,
  type EpisodeFileHint,
  type DebridServiceType,
  DebridServiceType as DebridServiceTypeNS,
  type DebridTorrent,
  type StreamInfo,
  AudioFormat,
  lastPathComponent,
  SourceType,
  VideoCodec,
  VideoQuality,
} from "./models";
import {
  type DebridService,
  DebridError,
  type FetchImpl,
  defaultFetchImpl,
  formValueEncode,
} from "./types";

/** Async delay used by retry/poll loops; default is a no-op so tests don't sleep. */
type Sleep = (ms: number) => Promise<void>;
const noopSleep: Sleep = () => Promise.resolve();

/** Coerces a JSON value to a number (mirrors the Swift `int64Value` helper that
 * accepts NSNumber / Int / numeric String). Returns null when not coercible. */
function int64Value(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string") {
    const n = Number.parseInt(value, 10);
    return Number.isNaN(n) ? null : n;
  }
  return null;
}

export class RealDebridService implements DebridService {
  readonly serviceType: DebridServiceType = DebridServiceTypeNS.realDebrid;
  private readonly apiToken: string;
  private readonly baseURL = "https://api.real-debrid.com/rest/1.0";
  private readonly fetchImpl: FetchImpl;
  private readonly sleep: Sleep;

  constructor(apiToken: string, fetchImpl?: FetchImpl, sleep: Sleep = noopSleep) {
    this.apiToken = apiToken;
    this.fetchImpl = fetchImpl ?? defaultFetchImpl();
    this.sleep = sleep;
  }

  // MARK: - Cache Check

  async checkCache(hashes: string[]): Promise<Record<string, CacheStatus>> {
    if (hashes.length === 0) return {};
    // RD disabled /torrents/instantAvailability - return .unknown for all.
    const results: Record<string, CacheStatus> = {};
    for (const hash of hashes) {
      results[hash.toLowerCase()] = CacheStatus.unknown;
    }
    return results;
  }

  // MARK: - Magnet Operations

  async addMagnet(hash: string): Promise<string> {
    const magnet = `magnet:?xt=urn:btih:${hash}`;
    const body = `magnet=${formValueEncode(magnet)}`;

    const maxRetries = 5;
    let lastError: unknown;
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const data = await this.requestRaw("/torrents/addMagnet", "POST", body);
        const json = parseJSON(data);
        const id = json?.id;
        if (typeof id === "string") return id;
        throw DebridError.downloadFailed("Failed to parse magnet response");
      } catch (error) {
        lastError = error;
        if (
          error instanceof DebridError &&
          error.kind === "httpError" &&
          error.statusCode != null &&
          error.statusCode >= 500 &&
          error.statusCode <= 599
        ) {
          await this.sleep(2 ** attempt * 1000);
          continue;
        }
        throw error;
      }
    }
    throw lastError instanceof Error
      ? lastError
      : DebridError.downloadFailed(`Failed to add magnet after ${maxRetries} retries`);
  }

  async selectFiles(torrentId: string, fileIds: number[]): Promise<void> {
    const files = fileIds.length === 0 ? "all" : fileIds.map(String).join(",");
    const body = `files=${files}`;
    // 204 -> empty body; a non-2xx is a real failure and surfaces.
    await this.requestRaw(`/torrents/selectFiles/${torrentId}`, "POST", body);
  }

  async getStreamURL(
    torrentId: string,
    fileHint: EpisodeFileHint | null = null,
  ): Promise<StreamInfo> {
    let status = "";
    let json: Record<string, unknown> = {};
    const backoffSchedule = [0.4, 0.8, 1.5, 3.0];
    const backoffCap = 5.0;
    const maxAttempts = 20;

    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const data = await this.requestRaw(`/torrents/info/${torrentId}`, "GET");
      const parsed = parseJSON(data);
      if (parsed == null) throw DebridError.torrentNotFound(torrentId);
      json = parsed;
      status = typeof json.status === "string" ? json.status : "";

      if (status === "downloaded") break;

      if (
        status === "error" ||
        status === "dead" ||
        status === "virus" ||
        status === "magnet_error"
      ) {
        throw DebridError.downloadFailed(`Torrent status: ${status}`);
      }

      if (status === "waiting_files_selection") {
        await this.selectFiles(torrentId, []);
      }

      if (attempt < maxAttempts - 1) {
        const delaySeconds =
          attempt < backoffSchedule.length ? backoffSchedule[attempt] : backoffCap;
        await this.sleep(delaySeconds * 1000);
      }
    }

    if (status !== "downloaded") {
      throw DebridError.downloadFailed(`Torrent not ready. Status: ${status}`);
    }

    const links = Array.isArray(json.links) ? (json.links as unknown[]) : null;
    const stringLinks =
      links?.filter((l): l is string => typeof l === "string") ?? [];
    if (stringLinks.length === 0) throw DebridError.noFilesAvailable();

    const candidates = this.fileCandidates(json, stringLinks);
    const selected = DebridFileSelector.selectBest(candidates, fileHint);
    if (selected == null) throw DebridError.noFilesAvailable();

    const unrestricted = await this.unrestrictDetailed(selected.link);

    const filename = lastPathComponent(selected.fileName);
    const bytes = selected.sizeBytes;

    return {
      streamURL: unrestricted.download,
      quality: VideoQuality.parse(filename),
      codec: VideoCodec.parse(filename),
      audio: AudioFormat.parse(filename),
      source: SourceType.parse(filename),
      sizeBytes: bytes,
      fileName: filename,
      debridService: "RD",
      // Carried for the in-window transcode path (`/streaming/transcode/{id}`).
      restrictedId: unrestricted.id ?? undefined,
    };
  }

  private fileCandidates(
    json: Record<string, unknown>,
    links: string[],
  ): DebridFileCandidate[] {
    const files = Array.isArray(json.files)
      ? (json.files as Record<string, unknown>[])
      : null;
    if (files) {
      const selectedFiles = files
        .filter((f) => int64Value(f.selected) === 1)
        .sort((lhs, rhs) => {
          const l = int64Value(lhs.id) ?? Number.MAX_SAFE_INTEGER;
          const r = int64Value(rhs.id) ?? Number.MAX_SAFE_INTEGER;
          return l - r;
        });

      const candidates: DebridFileCandidate[] = [];
      links.forEach((link, index) => {
        if (index < selectedFiles.length) {
          const file = selectedFiles[index];
          const path =
            (typeof file.path === "string" ? file.path : null) ??
            (typeof file.filename === "string" ? file.filename : null) ??
            "Unknown";
          const size = int64Value(file.bytes) ?? 0;
          candidates.push({ link, fileName: path, sizeBytes: size });
        }
      });
      if (candidates.length > 0) return candidates;
    }

    const fallbackName =
      typeof json.filename === "string" ? json.filename : "Unknown";
    const fallbackSize = int64Value(json.bytes) ?? 0;
    return links.map((link) => ({
      link,
      fileName: fallbackName,
      sizeBytes: fallbackSize,
    }));
  }

  // MARK: - Unrestrict

  async unrestrict(link: string): Promise<string> {
    return (await this.unrestrictDetailed(link)).download;
  }

  /** Like {@link unrestrict} but also surfaces Real-Debrid's unrestrict `id`
   * (the key for the `/streaming/transcode/{id}` and `/streaming/mediaInfos/{id}`
   * endpoints). The `download` URL is required; `id` is best-effort (null when
   * RD omits it). Same retry-on-5xx behavior as `unrestrict`. */
  async unrestrictDetailed(
    link: string,
  ): Promise<{ download: string; id: string | null }> {
    const body = `link=${formValueEncode(link)}`;

    const maxRetries = 5;
    let lastError: unknown;
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const data = await this.requestRaw("/unrestrict/link", "POST", body);
        const json = parseJSON(data);
        const downloadStr = json?.download;
        if (typeof downloadStr === "string" && isValidURL(downloadStr)) {
          const id = typeof json?.id === "string" ? json.id : null;
          return { download: downloadStr, id };
        }
        throw DebridError.downloadFailed("Failed to parse unrestrict response");
      } catch (error) {
        lastError = error;
        if (
          error instanceof DebridError &&
          error.kind === "httpError" &&
          error.statusCode != null &&
          error.statusCode >= 500 &&
          error.statusCode <= 599
        ) {
          await this.sleep(2 ** attempt * 1000);
          continue;
        }
        throw error;
      }
    }
    throw lastError instanceof Error
      ? lastError
      : DebridError.downloadFailed(`Failed to unrestrict link after ${maxRetries} retries`);
  }

  // MARK: - Streaming / Transcode

  /** Fetch Real-Debrid's transcoded HLS (`.m3u8`) URL for an unrestrict `id`.
   *
   * `GET /streaming/transcode/{id}` returns a map of streaming formats:
   *   { apple: {...}, dash: {...}, liveMP4: {...}, h264WebM: {...} }
   * where `apple` is the HLS (M3U8) format - a quality-keyed map (e.g. `full`,
   * `1080p`, `720p`, `480p`) whose values are `.m3u8` URLs. We prefer the
   * highest-quality variant (`full` if present, else the highest resolution,
   * else any HLS URL), so an MKV/HEVC source can be played in-webview by hls.js.
   *
   * Returns null when no HLS variant is available (the caller falls back to the
   * native-player hand-off). Defensive against shape drift: tolerates missing
   * keys, non-string values, and `apple` being absent. */
  async getTranscodeHLS(unrestrictId: string): Promise<string | null> {
    const data = await this.requestRaw(
      `/streaming/transcode/${unrestrictId}`,
      "GET",
    );
    const json = parseJSON(data);
    if (json == null) return null;
    return pickBestHLS(json);
  }

  /** Fetch Real-Debrid media info (`GET /streaming/mediaInfos/{id}`) for an
   * unrestrict `id` - codec/format/duration details. Returned as the raw parsed
   * object (the shape is large and only loosely documented); null on parse
   * failure. Currently informational; the transcode path keys off the filename
   * codec, but this lets callers double-check container/codec at resolve time. */
  async getMediaInfos(
    unrestrictId: string,
  ): Promise<Record<string, unknown> | null> {
    const data = await this.requestRaw(
      `/streaming/mediaInfos/${unrestrictId}`,
      "GET",
    );
    return parseJSON(data);
  }

  // MARK: - User Torrents

  /** Returns the torrent ID if a matching hash is found (downloaded/in-progress),
   * deletes error-state matches, and returns null otherwise. */
  async findExistingTorrent(hash: string): Promise<string | null> {
    const lowerHash = hash.toLowerCase();
    const pageSize = 100;
    const maxPages = 20;
    // Paginate like listTorrents - a single page=1 fetch silently misses an
    // already-downloaded torrent on accounts with >100 torrents, forcing a
    // duplicate re-add instead of instantly reusing the cached one.
    for (let page = 1; page <= maxPages; page += 1) {
      const data = await this.requestRaw(
        `/torrents?limit=${pageSize}&page=${page}`,
        "GET",
      );
      const torrents = parseJSONArray(data);
      if (torrents == null) return null;

      for (const torrent of torrents) {
        const torrentHash = torrent.hash;
        const id = torrent.id;
        if (
          typeof torrentHash === "string" &&
          torrentHash.toLowerCase() === lowerHash &&
          typeof id === "string"
        ) {
          const status = typeof torrent.status === "string" ? torrent.status : "";
          if (status === "downloaded") return id;
          if (status === "error" || status === "dead" || status === "magnet_error") {
            try {
              await this.deleteTorrent(id);
            } catch {
              // best-effort delete (mirrors Swift `try?`)
            }
            return null;
          }
          return id;
        }
      }
      if (torrents.length < pageSize) break; // last (short/empty) page
    }
    return null;
  }

  /** Delete a torrent from the user's list. */
  async deleteTorrent(id: string): Promise<void> {
    await this.requestRaw(`/torrents/delete/${id}`, "DELETE");
  }

  /** List the account's torrents (the Debrid Library manager source). Walks the
   * paginated `GET /torrents?limit=&page=` endpoint until a short/empty page is
   * returned (or `maxPages` is hit), normalizing each into a `DebridTorrent`.
   * Fault-tolerant: a page that fails to parse stops pagination rather than
   * throwing, so a partial list still surfaces. */
  async listTorrents(maxPages = 20, pageSize = 100): Promise<DebridTorrent[]> {
    const out: DebridTorrent[] = [];
    for (let page = 1; page <= maxPages; page++) {
      const data = await this.requestRaw(
        `/torrents?limit=${pageSize}&page=${page}`,
        "GET",
      );
      const torrents = parseJSONArray(data);
      if (torrents == null || torrents.length === 0) break;
      for (const t of torrents) {
        out.push(normalizeRDTorrent(t));
      }
      if (torrents.length < pageSize) break;
    }
    return out;
  }

  // MARK: - Account

  async validateToken(): Promise<boolean> {
    try {
      await this.getAccountInfo();
      return true;
    } catch {
      return false;
    }
  }

  async getAccountInfo(): Promise<DebridAccountInfo> {
    const data = await this.requestRaw("/user", "GET");
    const json = parseJSON(data);
    if (json == null) throw DebridError.invalidToken();

    const username = typeof json.username === "string" ? json.username : "Unknown";
    const email = typeof json.email === "string" ? json.email : null;
    const premium = typeof json.premium === "number" ? json.premium : 0;
    const points = typeof json.points === "number" ? json.points : null;

    let premiumExpiry: Date | null = null;
    if (typeof json.expiration === "string") {
      const parsed = new Date(json.expiration);
      premiumExpiry = Number.isNaN(parsed.getTime()) ? null : parsed;
    }

    return {
      username,
      email,
      premiumExpiry,
      isPremium: premium > 0,
      points,
    };
  }

  // MARK: - HTTP

  private async requestRaw(
    path: string,
    method: string,
    body?: string,
  ): Promise<string> {
    const url = this.baseURL + path;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiToken}`,
    };
    if (body != null) {
      headers["Content-Type"] = "application/x-www-form-urlencoded";
    }

    const response = await this.fetchImpl(url, { method, headers, body });
    const status = response.status;

    // 204 No Content is success (selectFiles, delete).
    if (status === 204) return "";

    if (!(status >= 200 && status <= 299)) {
      if (status === 401) throw DebridError.invalidToken();
      if (status === 403) throw DebridError.expired();
      if (status === 429) throw DebridError.rateLimited();
      const errorMsg = (await response.text().catch(() => "")) || "Unknown error";
      throw DebridError.httpError(status, errorMsg);
    }

    return response.text();
  }
}

// MARK: - JSON helpers

function parseJSON(text: string): Record<string, unknown> | null {
  if (text.length === 0) return null;
  try {
    const value = JSON.parse(text);
    return value != null && typeof value === "object" && !Array.isArray(value)
      ? (value as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}

function parseJSONArray(text: string): Record<string, unknown>[] | null {
  if (text.length === 0) return null;
  try {
    const value = JSON.parse(text);
    return Array.isArray(value) ? (value as Record<string, unknown>[]) : null;
  } catch {
    return null;
  }
}

function isValidURL(value: string): boolean {
  try {
    // Mirrors `URL(string:)` - requires an absolute URL.
    new URL(value);
    return true;
  } catch {
    return false;
  }
}

/** Normalize a raw Real-Debrid `/torrents` row into the shared `DebridTorrent`
 * display shape. Defensive against missing/typed-wrong fields. */
function normalizeRDTorrent(t: Record<string, unknown>): DebridTorrent {
  const id = typeof t.id === "string" ? t.id : String(t.id ?? "");
  const name = typeof t.filename === "string" ? t.filename : "Unknown";
  const hash = typeof t.hash === "string" ? t.hash.toLowerCase() : null;
  const status = typeof t.status === "string" ? t.status : "unknown";
  const host = typeof t.host === "string" ? t.host : null;
  const added = typeof t.added === "string" ? t.added : null;
  const bytes = int64Value(t.bytes) ?? 0;
  const progress = typeof t.progress === "number" ? t.progress : null;
  return {
    id,
    name,
    sizeBytes: bytes,
    status,
    infoHash: hash,
    addedAt: added,
    host,
    progress,
    debridService: "RD",
  };
}

/** Rank a transcode quality key so we can prefer the highest variant. RD's
 * `apple` map is keyed by labels like `full`, `1080p`, `720p`, `480p`. `full`
 * is the original-resolution stream and wins; otherwise higher pixel height
 * wins; unknown labels sort last. */
function hlsQualityRank(key: string): number {
  const k = key.toLowerCase();
  if (k === "full" || k === "original" || k === "max") return 100000;
  const res = k.match(/(\d{3,4})\s*p?/);
  if (res) return Number.parseInt(res[1], 10);
  return -1;
}

/** Extract the best HLS (`.m3u8`) URL from a `/streaming/transcode/{id}`
 * response. Prefers the `apple` (HLS) format's highest-quality variant. Falls
 * back to scanning any nested string that looks like an `.m3u8` URL so a minor
 * shape change (e.g. the HLS map under a different key) still resolves. Returns
 * null when nothing usable is found. */
function pickBestHLS(json: Record<string, unknown>): string | null {
  const isM3U8 = (v: unknown): v is string =>
    typeof v === "string" && isValidURL(v) && v.split("?")[0].toLowerCase().includes(".m3u8");

  // Primary: the documented `apple` (M3U8) quality map.
  const apple = json.apple;
  if (apple != null && typeof apple === "object" && !Array.isArray(apple)) {
    let best: { url: string; rank: number } | null = null;
    for (const [key, value] of Object.entries(apple as Record<string, unknown>)) {
      if (!isM3U8(value)) continue;
      const rank = hlsQualityRank(key);
      if (best == null || rank > best.rank) best = { url: value, rank };
    }
    if (best != null) return best.url;
  }

  // Defensive fallback: any `.m3u8` URL anywhere in the response object. Walks
  // one or two levels (top-level strings + nested quality maps).
  for (const value of Object.values(json)) {
    if (isM3U8(value)) return value;
    if (value != null && typeof value === "object" && !Array.isArray(value)) {
      for (const nested of Object.values(value as Record<string, unknown>)) {
        if (isM3U8(nested)) return nested;
      }
    }
  }
  return null;
}
